pkg/util/labels.go

```diff
@@ -90,9 +113,75 @@ func AddObjectLabels(obj runtime.Object, labels labels.Set) error {
 		return addDeploymentConfigNestedLabels(objType, labels)
 	}
 
+	// TODO: allow meta.Accessor to handle runtime.Unstructured
+	if unstruct, ok := obj.(*runtime.Unstructured); ok && unstruct.Object != nil {
+		// the presence of "metadata" is sufficient for us to apply the rules for Kube-like
+		// objects.
+		// TODO: add swagger detection to allow this to happen more effectively
+		if obj, ok := unstruct.Object["metadata"]; ok {
+			if m, ok := obj.(map[string]interface{}); ok {
+
+				existing := make(map[string]string)
+				if l, ok := m["labels"]; ok {
+					if found, ok := extractLabels(l); ok {
+						existing = found
+					}
+				}
+				if err := MergeInto(existing, labels, OverwriteExistingDstKey); err != nil {
+					return err
+				}
+				m["labels"] = mapToGeneric(existing)
+			}
+		}
+
+		// only attempt to set root labels if a root object called labels exists
+		// TODO: add swagger detection to allow this to happen more effectively
+		if obj, ok := unstruct.Object["labels"]; ok {
+			existing := make(map[string]string)
+			if found, ok := extractLabels(obj); ok {
+				existing = found
+			}
+			if err := MergeInto(existing, labels, OverwriteExistingDstKey); err != nil {
+				return err
+			}
+			unstruct.Object["labels"] = mapToGeneric(existing)
+		}
+
+		// handle nested labels in unstructured objects
+		return addUnstructuredNestedLabels(unstruct, labels)
+	}
+
 	return nil
 }
 
+// extractLabels extracts a map[string]string from a map[string]interface{}
+func extractLabels(obj interface{}) (map[string]string, bool) {
```

Comment:
Maybe we should add a nil check of this `obj` just like `mapToGeneric` is doing.